File: Pablo/Examenes/Febrero2019/feb2019java/src/Kruskal.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:
 * -- GRADO/STUDIES:
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

import dataStructures.graph.WeightedGraph;
import dataStructures.graph.WeightedGraph.WeightedEdge;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.priorityQueue.LinkedPriorityQueue;
import dataStructures.set.Set;
import dataStructures.set.HashSet;
import dataStructures.tuple.Tuple2;

import java.util.Iterator;
import java.util.Stack;

public class Kruskal {
	public static <V,W> Set<WeightedEdge<V,W>> kruskal(WeightedGraph<V,W> g) {
		Dictionary<V,V> DICT = new HashDictionary<>();
		for (V v : g.vertices())
		{
			DICT.insert(v, v);
		}

		PriorityQueue<WeightedEdge<V,W>> PQ = new LinkedPriorityQueue<>();
		Iterator<WeightedEdge<V,W>> it = g.edges().iterator();
		while(it.hasNext()){
			PQ.enqueue(it.next());
		}

		Set<WeightedEdge<V,W>> T = new HashSet<>();

		while(!PQ.isEmpty())
		{
			WeightedEdge<V,W> edge = PQ.first();
			PQ.dequeue();
			V src = edge.source();
			V rep = DICT.valueOf(src);
			while(src != rep){
				src = rep;
				rep = DICT.valueOf(src);
			}

			V src2 = edge.destination();
			V rep2 = DICT.valueOf(src2);
			while(src2 != rep2){
				src2 = rep2;
				rep2 = DICT.valueOf(src2);
			}

			if(rep != rep2) {
				DICT.delete(rep2);
				DICT.insert(rep2, edge.source());
				T.insert(edge);
			}
		}

		return T;
	}

	// Sólo para evaluación continua / only for part time students
	public static <V,W> Set<Set<WeightedEdge<V,W>>> kruskals(WeightedGraph<V,W> g) {

		// COMPLETAR
		
		return null;
	}
}

File: Pablo/Examenes/Febrero2019/feb2019haskell/Kruskal.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
----------------------------------------------

module Kruskal(kruskal, kruskals) where

import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.LinearPriorityQueue as Q
import DataStructures.Graph.DictionaryWeightedGraph

kruskal :: (Ord a, Ord w) => WeightedGraph a w -> [WeightedEdge a w]
kruskal dic = aux (initDic (vertices dic)) (initPQ (edges dic)) []
    where 
        initDic [] = D.empty
        initDic (x:xs) = D.insert x x (initDic xs)

        initPQ [] = Q.empty
        initPQ (x:xs) = Q.enqueue x (initPQ xs)

        aux dict pq sol = if(Q.isEmpty pq) then sol else (solucionar dict pq sol)
            where 
                (WE a w b) = Q.first pq
                representante src dic = case D.valueOf src dic of
                    Just val -> if(val == src) then src else representante val dic
                    Nothing -> error "something has gone wrong"
                solucionar dict pq sol = if((representante a dict) /= (representante b dict)) then ((WE a w b) : 
                    (aux (modDic dict) (Q.dequeue pq) sol)) else aux dict (Q.dequeue pq) sol
                modDic dic = D.insert a (representante b dict) (D.delete a dic) 


-- Solo para evaluación continua / only for part time students
kruskals :: (Ord a, Ord w) => WeightedGraph a w -> [[WeightedEdge a w]]
kruskals = undefined
